packages/NodeTypeResolver/PHPStan/Scope/NodeVisitor/ByRefVariableNodeVisitor.php

```diff
@@ -0,0 +1,71 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Rector\NodeTypeResolver\PHPStan\Scope\NodeVisitor;
+
+use PhpParser\Node;
+use PhpParser\Node\Expr\AssignRef;
+use PhpParser\Node\Expr\Closure;
+use PhpParser\Node\Expr\Variable;
+use PhpParser\Node\FunctionLike;
+use PhpParser\NodeVisitorAbstract;
+use Rector\NodeTypeResolver\Node\AttributeKey;
+use Rector\PhpDocParser\NodeTraverser\SimpleCallableNodeTraverser;
+
+final class ByRefVariableNodeVisitor extends NodeVisitorAbstract
+{
+    public function __construct(
+        private readonly SimpleCallableNodeTraverser $simpleCallableNodeTraverser
+    ) {
+    }
+
+    public function enterNode(Node $node): ?Node
+    {
+        if (! $node instanceof FunctionLike) {
+            return null;
+        }
+
+        $stmts = $node->getStmts();
```

Comment:
I wonder why a visitor needs another visitor to visit the statements?

this visitor will visit the statements automatically because of the recursive nature?

Reply:
that's checked above it that limit what stmts to be checked.

```php
        if (! $node instanceof FunctionLike) {
            return null;
        }
```

it only check the Node of stmt per visit that a `FunctionLike` instance.